File: app/src/main/java/tabletop/velocic/com/worldforgerpgtools/GeneratorDeserializer/GeneratorImporter.java
package tabletop.velocic.com.worldforgerpgtools.GeneratorDeserializer;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.AssetManager;
import android.util.Base64;
import android.util.Log;

import com.google.gson.Gson;
import com.google.gson.JsonSyntaxException;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

public class GeneratorImporter
{
    public static final String TAG_GENERATOR_IMPORT = "GENERATOR IMPORT";
    public static final String GENERATOR_DATA_FOLDER = "GeneratorData";
    public static final String IMPORTER_PREFERENCES_FILE = "GeneratorImporterPrefs";

    private static GeneratorImporter importerInstance;
    private GeneratorCategory rootGeneratorCategory;

    private GeneratorImporter() {}

    public static GeneratorImporter getInstance(Context context)
    {
        if (importerInstance == null) {
            importerInstance = new GeneratorImporter();

            AssetManager assetManager = context.getAssets();
            SharedPreferences sharedPrefs = context.getSharedPreferences(IMPORTER_PREFERENCES_FILE, Context.MODE_PRIVATE);
            SharedPreferences.Editor sharedPrefsEditor = sharedPrefs.edit();
            importerInstance.oneTimeLocalStorageCopy(context, assetManager, sharedPrefs, sharedPrefsEditor, GENERATOR_DATA_FOLDER);
            sharedPrefsEditor.apply();

            importerInstance.importGenerators(context);
        }

        return importerInstance;
    }

    private void oneTimeLocalStorageCopy(Context context, AssetManager assets, SharedPreferences sharedPrefs, SharedPreferences.Editor sharedPrefsEditor, String currentPath)
    {
        try {
            String[] contents = assets.list(currentPath);
            MessageDigest md5 = MessageDigest.getInstance("MD5");

            //Path is a non-empty directory
            if (contents.length > 0) {
                byte[] hash = md5.digest(currentPath.getBytes());
                String encodedHash = Base64.encodeToString(hash, Base64.DEFAULT);
                boolean hasBeenCopied = sharedPrefs.getBoolean(encodedHash, false);

                int extensionIndex = currentPath.lastIndexOf(".");

                String extension = "";

                if (extensionIndex != -1) {
                    extension = currentPath.substring(extensionIndex + 1);
                }

                if (hasBeenCopied == false && extension == "") {
                    sharedPrefsEditor.putBoolean(encodedHash, true);

                    String subDirPathString = context.getFilesDir() + "/" + currentPath;

                    //Create subdirectory (getDir can only create at the top-level)
                    File subDirectory = new File(subDirPathString);
                    subDirectory.mkdir();
                }

                //TODO: Don't traverse further if the directory doesn't exist and we didn't add it

                for (String item : contents) {
                    oneTimeLocalStorageCopy(context, assets, sharedPrefs, sharedPrefsEditor, currentPath + "/" + item);
                }
            } else {
                //Path is a file

                //Read the source file from assets & hash the file contents
                InputStream sourceStream = assets.open(currentPath);
                String encodedHash = createChecksum(sourceStream, "MD5");
                sourceStream.close();

                boolean hasBeenCopied = sharedPrefs.getBoolean(encodedHash, false);

                if (hasBeenCopied == false) {
                    sharedPrefsEditor.putBoolean(encodedHash, true);

                    //Reopen the stream to copy the file to the new destination
                    sourceStream = assets.open(currentPath);

                    //Create a new file in internal storage
                    String newFilePath = context.getFilesDir() + "/" + currentPath;
                    File file = new File(newFilePath);
                    FileOutputStream outputStream;

                    byte[] buffer = new byte[32768];

                    if (file.exists() == false) {
                        file.createNewFile();
                    }

                    outputStream = new FileOutputStream(file);

                    int bytesRead = 0;
                    while ((bytesRead = sourceStream.read(buffer)) > 0) {
                        outputStream.write(buffer, 0, bytesRead);
                    }

                    outputStream.close();
                }
            }
        } catch (IOException e) {
            Log.d(TAG_GENERATOR_IMPORT, "Failed to one-time copy pre-made tables from assets directory to internal storage: " + e.getMessage());
        } catch (NoSuchAlgorithmException e) {
            Log.d(TAG_GENERATOR_IMPORT, e.getMessage());
        }
    }

    private String createChecksum(InputStream input, String digestType)
    {
        MessageDigest digest = null;
        String checksum = "";
        int numRead = 0;

        try {
            digest = MessageDigest.getInstance(digestType);
        } catch (NoSuchAlgorithmException e) {
            Log.d(TAG_GENERATOR_IMPORT, e.getMessage());
        }

        if (digest == null) {
            return checksum;
        }

        byte[] buffer = new byte[32768];

        try {
            do {
                numRead = input.read(buffer);
                if (numRead > 0) {
                    digest.update(buffer, 0, numRead);
                }
            } while (numRead != -1);

        } catch (IOException e) {
            Log.d(TAG_GENERATOR_IMPORT, "Failed to calculate checksum: " + e.getMessage());
        }

        checksum = Base64.encodeToString(digest.digest(), Base64.DEFAULT);

        return checksum;
    }

    public void importGenerators(Context context)
    {
        AssetManager assetManager = context.getAssets();

        rootGeneratorCategory = loadGeneratorCategories(
            new GeneratorCategory("root", GENERATOR_DATA_FOLDER),
            GENERATOR_DATA_FOLDER,
            assetManager
        );

        rootGeneratorCategory = populateGenerators(rootGeneratorCategory, assetManager);
    }

    private GeneratorCategory loadGeneratorCategories(GeneratorCategory parent, File internalStorageDirectory)
    {
        File[] items = internalStorageDirectory.listFiles();

        if (items.length > 0) {
            for (File item : items) {
                String fullPath = item.getAbsolutePath();
                String fileExtension = fullPath.substring(fullPath.lastIndexOf(".") + 1);

                if (item.isDirectory()) {
                    if (fileExtension.equals("json")) {
                        parent.addGeneratorJsonDataPath(fullPath);
                        continue;
                    }

                    GeneratorCategory childCategory = new GeneratorCategory(item.getName(), fullPath, parent);
                    parent.addChildCategory(loadGeneratorCategories(childCategory, item));
                }
            }

            return parent;
        }

        String fullPath = internalStorageDirectory.getAbsolutePath();
        String fileExtension = fullPath.substring(fullPath.lastIndexOf(".") + 1);

        if (fileExtension.isEmpty()) {
            String categoryName = fullPath.substring(fullPath.lastIndexOf(".") + 1);
            parent.addChildCategory(new GeneratorCategory(categoryName, fullPath, parent));
        }

        return parent;
    }

    private GeneratorCategory loadGeneratorCategories(GeneratorCategory parent, String path, AssetManager assets)
    {
        try {
            String[] contents = assets.list(path);

            //Path is a non-empty directory
            if (contents.length > 0) {
                for (String item : contents) {
                    String fileExtension = item.substring(item.lastIndexOf(".") + 1);
                    String fullPath = path + "/" + item;

                    if (fileExtension.equals("json")) {
                        parent.addGeneratorJsonDataPath(fullPath);
                        continue;
                    }

                    GeneratorCategory childCategory = new GeneratorCategory(item, fullPath, parent);
                    parent.addChildCategory(loadGeneratorCategories(childCategory, fullPath, assets));
                }

                return parent;
            }

            //Path is a file (non-json, which shouldn't be here), or an empty directory. Check extension
            //to try and prevent adding invalid files as categories.
            String extension = path.substring(path.lastIndexOf(".") + 1);
            if (extension.isEmpty()) {
                String categoryName = path.substring(path.lastIndexOf("/" + 1));
                parent.addChildCategory(new GeneratorCategory(categoryName, path, parent));
            }
            return parent;

        } catch (IOException e) {
            Log.d(TAG_GENERATOR_IMPORT, "Failed to import random generator tables: " + e.getMessage());
            return parent;
        }
    }

    private GeneratorCategory populateGenerators(GeneratorCategory rootNode, AssetManager assets)
    {
        //TODO: Later optimization, lazy load JSON files as needed if loading times or
        //memory usage is too much.
        Gson gson = new Gson();
        for (GeneratorCategory child : rootNode.getChildCategories()) {
            for (String jsonDataPath : child.getGeneratorJsonDataPaths()) {
                try {
                    InputStream jsonInputStream = assets.open(jsonDataPath);
                    Reader jsonInputReader = new InputStreamReader(jsonInputStream);

                    Generator generator = gson.fromJson(jsonInputReader, Generator.class);
                    generator.setAssetPath(jsonDataPath);
                    child.addGenerator(generator);
                } catch (IOException e) {
                    Log.d(TAG_GENERATOR_IMPORT, "Failed to deserialize " + jsonDataPath + ": " + e.getMessage());
                    continue;
                } catch (JsonSyntaxException e) {
                    Log.d(TAG_GENERATOR_IMPORT, "Invalid JSON syntax in " + jsonDataPath + ": " + e.getMessage());
                    continue;
                }
            }

            populateGenerators(child, assets);
        }

        return rootNode;
    }

    public GeneratorCategory getRootGeneratorCategory()
    {
        return rootGeneratorCategory;
    }
}

File: app/src/main/java/tabletop/velocic/com/worldforgerpgtools/GeneratorDeserializer/GeneratorImporter.kt
package tabletop.velocic.com.worldforgerpgtools.GeneratorDeserializer

import android.content.Context
import android.content.SharedPreferences
import android.content.res.AssetManager
import android.util.Base64
import android.util.Log
import com.google.gson.Gson
import com.google.gson.JsonSyntaxException
import java.io.*
import java.security.MessageDigest

object GeneratorImporter {
    const val GENERATOR_DATA_FOLDER = "GeneratorData"
    private const val IMPORTER_PREFERENCES_FILE = "GeneratorImporterPrefs"
    private const val TAG_GENERATOR_IMPORT = "GENERATOR IMPORT"
    private const val BUFFER_SIZE = 32768

    @JvmStatic
    var rootGeneratorCategory: GeneratorCategory? = null
        private set

    fun import(context: Context?) {

        if (context == null) {
            Log.d(TAG_GENERATOR_IMPORT, "Received a null context while attempting to import data from the filesystem; aborting.")
            return
        }

        val assetManager = context.assets
        val sharedPrefs = context.getSharedPreferences(IMPORTER_PREFERENCES_FILE, Context.MODE_PRIVATE)
        val prefsEditor = sharedPrefs.edit()
        oneTimeLocalStorageCopy(context, assetManager, sharedPrefs, prefsEditor, GENERATOR_DATA_FOLDER)
        prefsEditor.apply()

        importGenerators(context)
    }

    private fun oneTimeLocalStorageCopy(
        context: Context,
        assetManager: AssetManager,
        sharedPrefs: SharedPreferences,
        prefsEditor: SharedPreferences.Editor,
        currentPath: String
    ) {
        val contents = assetManager.list(currentPath)
        val messageDigest = MessageDigest.getInstance("MD5")

        if (contents.isNotEmpty()) {
            val encodedHash = messageDigest.digest(currentPath.toByteArray()).toString()
            val hasBeenCopied = sharedPrefs.getBoolean(encodedHash, false)

            val extension = currentPath.let {
                val extensionIndex = it.lastIndexOf('.')
                if (extensionIndex != -1) it.substring(extensionIndex + 1) else ""
            }

            //Target to copy is a directory (all data files can be assumed to have an extension in this app)
            if (!hasBeenCopied && extension == "") {
                prefsEditor.putBoolean(encodedHash, true)
                val subDirPathString = "${context.filesDir}/$currentPath"

                try {
                    File(subDirPathString).mkdir()
                } catch (e: IOException) {
                    Log.d(TAG_GENERATOR_IMPORT, "Failed to one-time copy premade tables from assets directory to internal storage: ${e.message}")
                    return
                }
            }

            //TODO: Don't traverse further if the directory doesn't exist and we didn't add it
            contents.forEach {
                oneTimeLocalStorageCopy(context, assetManager, sharedPrefs, prefsEditor, "$currentPath/$it")
            }
        } else {
            //Path is a file

            //Read the source file from assets & hash the file contents
            var sourceStream = assetManager.open(currentPath)
            val encodedHash = createChecksum(sourceStream, "MD5")
            val hasBeenCopied = sharedPrefs.getBoolean(encodedHash, true)

            if (!hasBeenCopied) {
                prefsEditor.putBoolean(encodedHash, true)

                //Reopen the stream to copy the file to the new destination
                sourceStream = assetManager.open(currentPath)

                //Create a new file in internal storage
                val file = File("${context.filesDir}/$currentPath")
                val buffer = ByteArray(BUFFER_SIZE)

                if (!file.exists()) {
                    try {
                        file.createNewFile()
                    } catch (e: IOException) {
                        Log.d(TAG_GENERATOR_IMPORT, "Failed to copy $currentPath from assets directory into internal storage: ${e.message}")
                    }
                }

                val outputStream = FileOutputStream(file)
                var bytesRead: Int

                do {
                    bytesRead = sourceStream.read(buffer)
                    outputStream.write(buffer, 0, bytesRead)
                } while (bytesRead > 0)

                outputStream.close()
            }
        }
    }

    private fun importGenerators(context: Context) {
        val assetManager = context.assets


        val rootGeneratorCategory = loadGeneratorCategories(
            GeneratorCategory("root", GENERATOR_DATA_FOLDER),
            File("${context.filesDir}/$GENERATOR_DATA_FOLDER")
        )

        this.rootGeneratorCategory = populateGenerators(rootGeneratorCategory, assetManager)
    }

    private fun loadGeneratorCategories(parent: GeneratorCategory, internalStorageTarget: File) : GeneratorCategory {
        val files = internalStorageTarget.listFiles()

        if (files.isNotEmpty()) {
            for (file in files) {
                //Truncate system directories out of the path. Full path begins from the contextual
                //root of our data files (GENERATOR_DATA_FOLDER)
                val rawPath = file.absolutePath
                val generatorDataIndex = rawPath.indexOf(GENERATOR_DATA_FOLDER)

                val fullPath = rawPath.substring(generatorDataIndex)
                val fileExtension = file.extension

                if (fileExtension == "json") {
                    parent.generatorJsonDataPaths.add(fullPath)
                    continue;
                }

                val childCategory = GeneratorCategory(file.name, fullPath, parent)
                parent.childCategories.add(loadGeneratorCategories(childCategory, file))
            }

            return parent
        }

        val rawPath = internalStorageTarget.absolutePath
        val generatorDataIndex = rawPath.indexOf(GENERATOR_DATA_FOLDER)

        val fullPath = rawPath.substring(generatorDataIndex)
        val fileExtension = internalStorageTarget.extension

        if (fileExtension.isEmpty()) {
            val categoryName = fullPath.substringAfterLast(".")
            parent.childCategories.add(GeneratorCategory(categoryName, fullPath, parent))
        }

        return parent
    }

    private fun populateGenerators(rootNode: GeneratorCategory, assets: AssetManager) : GeneratorCategory {
        val gson = Gson()

        for (child in rootNode.childCategories) {
            for (jsonDataPath in child.generatorJsonDataPaths) {
                try {
                    val jsonInputReader = InputStreamReader(assets.open(jsonDataPath))
                    val generator = gson.fromJson(jsonInputReader, Generator::class.java)

                    generator.assetPath = jsonDataPath
                    child.generators.add(generator)
                } catch (e: IOException) {
                    Log.d(TAG_GENERATOR_IMPORT, "Failed to deserialize $jsonDataPath: ${e.message}")
                    continue
                } catch (e: JsonSyntaxException) {
                    Log.d(TAG_GENERATOR_IMPORT, "Invalid JSON syntax in $jsonDataPath: ${e.message}")
                    continue
                }
            }

            populateGenerators(child, assets)
        }

        return rootNode
    }

    private fun createChecksum(input: InputStream, digestType: String) : String {
        val digest = MessageDigest.getInstance(digestType) ?: return ""
        var numBytesRead: Int
        val buffer = ByteArray(BUFFER_SIZE)

        do {
            numBytesRead = input.read(buffer)

            if (numBytesRead > 0) {
                digest.update(buffer, 0, numBytesRead)
            }
        } while (numBytesRead != -1)

        return Base64.encodeToString(digest.digest(), Base64.DEFAULT)
    }
}
